features/cryptocell/FEATURE_CRYPTOCELL310/aes_alt.c

```diff
@@ -0,0 +1,329 @@
+     }
+
+     ret = SaSi_AesFinish( &ctx->CC_Context, length,
+                           ( unsigned char* )input, length, output, &length );
+     if( ret != 0 )
+         return ( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+
+     /* update the IV for next block
+      * For CTR mode, update the nonce only if the current length is a full AES block length
+      */
+
+     if( ( ( aes_mode == SASI_AES_MODE_CBC ) ||
+          ( (aes_mode == SASI_AES_MODE_CTR) && ( ( length % SASI_AES_BLOCK_SIZE_IN_BYTES) == 0) ) )
+             && iv )
+     {
+          ret = SaSi_AesGetIv( &ctx->CC_Context, iv );
+          if( ret != 0 )
+              return ( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+      }
+
+      ret = SaSi_AesFree( &ctx->CC_Context );
+      if ( ret != 0 )
+          return ( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+
+      return ( 0 );
+}
+
+int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
+                    int mode,
+                    const unsigned char input[16],
+                    unsigned char output[16] )
+{
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    if( ( mode == MBEDTLS_AES_ENCRYPT  && ctx->CC_cipherFlag !=  SASI_AES_ENCRYPT ) ||
+        ( mode == MBEDTLS_AES_DECRYPT && ctx->CC_cipherFlag !=  SASI_AES_DECRYPT ) )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    return( CC_aes_cipher( ctx, mode, SASI_AES_MODE_ECB, 16, NULL, 0, input, output ) );
+
+}
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC)
+int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                           int mode,
+                           size_t length,
+                           unsigned char iv[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    if( ctx == NULL )
+        return ( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+
+    if( length % SASI_AES_BLOCK_SIZE_IN_BYTES )
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+
+    if( ( mode != MBEDTLS_AES_ENCRYPT  || ctx->CC_cipherFlag !=  SASI_AES_ENCRYPT ) &&
+        ( mode != MBEDTLS_AES_DECRYPT || ctx->CC_cipherFlag !=  SASI_AES_DECRYPT ) )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    return( CC_aes_cipher( ctx, mode, SASI_AES_MODE_CBC, length, iv, 16, input, output ) );
+}
+
+#endif /* MBEDTLS_CIPHER_MODE_CBC */
+
+#if defined(MBEDTLS_CIPHER_MODE_CTR)
+int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
+                           size_t length,
+                           size_t *nc_off,
+                           unsigned char nonce_counter[16],
+                           unsigned char stream_block[16],
+                           const unsigned char *input,
+                           unsigned char *output )
+{
+    int ret = 0;
+    int n = *nc_off, c, i;
+    size_t j;
+    if( ctx == NULL )
+        return( MBEDTLS_ERR_AES_BAD_INPUT_DATA );
+
+    if ( *nc_off )
+    {
+        /* handle corner case where we are resuming a previous encryption,
```

Comment:
Does CC_aes_cipher not handle this corner case itself?

Reply:
no. CC CTR implementation only supports where `*nc_off` is zero

Reply:
What happens if it isn't? Incorrect computation? How do users of the CC API cope with this limitation ordinarily?

Reply:
yes, incorrect computation. Users of CC, that are interst in `*nc_off ~= 0` will need to do the same thing